corehq/apps/linked_domain/updates.py

```diff
@@ -401,6 +413,94 @@ def update_hmac_callout_settings(domain_link):
+            new_rule = AutomaticUpdateRule(
+                domain=domain_link.linked_domain,
+                active=rule['rule']['active'],
+                workflow=AutomaticUpdateRule.WORKFLOW_CASE_UPDATE,
+            )
+
+            new_rule.name = rule['rule']['name']
+            new_rule.case_type = rule['rule']['case_type']
+            new_rule.filter_on_server_modified = rule['rule']['filter_on_server_modified']
+            new_rule.server_modified_boundary = rule['rule']['server_modified_boundary']
+            new_rule.save()
+
+            # Largely from data_interfacees/forms.py - save_criteria()
+            for criteria in rule['criteria']:
+                definition = None
+
+                if criteria['match_property_definition']:
+                    definition = MatchPropertyDefinition.objects.create(
+                        property_name=criteria['match_property_definition']['property_name'],
+                        property_value=criteria['match_property_definition']['property_value'],
+                        match_type=criteria['match_property_definition']['match_type'],
+                    )
+                elif criteria['custom_match_definition']:
+                    definition = CustomMatchDefinition.objects.create(
+                        name=criteria['custom_match_definition']['name'],
+                    )
+                elif criteria['closed_parent_definition']:
+                    definition = ClosedParentDefinition.objects.create()
+
+                new_criteria = CaseRuleCriteria(rule=new_rule)
+                new_criteria.definition = definition
+                new_criteria.save()
+
+            # Largely from data_interfacees/forms.py - save_actions()
+            for action in rule['actions']:
+                definition = None
+
+                if action['update_case_definition']:
+                    definition = UpdateCaseDefinition(close_case=action['update_case_definition']['close_case'])
+                    properties = []
+                    for propertyItem in action['update_case_definition']['properties_to_update']:
+                        properties.append(
+                            UpdateCaseDefinition.PropertyDefinition(
+                                name=propertyItem['name'],
+                                value_type=propertyItem['value_type'],
+                                value=propertyItem['value'],
+                            )
+                        )
+                    definition.set_properties_to_update(properties)
+                    definition.save()
+                elif action['custom_action_definition']:
+                    definition = CustomActionDefinition.objects.create(
+                        name=action['custom_action_definition']['name'],
+                    )
+
+                action = CaseRuleAction(rule=new_rule)
+                action.definition = definition
+                action.save()
+
+    for local_rule in local_rules:
```

Comment:
> Any rules that were created directly in the downstream domain, not linked from upstream, will be deleted. This could be okay, but it's inconsistent with how other linkable data models work.

Chiming in late here, but just want to add context. We should favor the solution that avoids clobbering rules created in the downstream domain. We (SaaS) want the downstream domain to be able to function as a typical domain (e.g., support app building without worrying about dependent data models being overridden from the upstream domain). We call these "destructive pushes" and there are a few other data models that behave this way, but as Jenny mentioned that is not the norm.

Looks like you've gone with the upstream_id approach though so this shouldn't be an issue, just wanted to mention this.

Reply:
Thanks for double-checking on that, Graham. You're right, Jenny noted this and I made changes to the update method so no downstream rules are deleted and thus there is no "destructive push".